### ISSUE TITLE
Perl Net-SSLeay testsuite failure in X509_CRL_digest

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#2313]
I've just been pointed to:
https://rt.cpan.org/Public/Bug/Display.html?id=120006 and https://bugs.debian.org/852920

Can someone take a look at this? @levitte?

**Comment by levitte** [comment]
Okei

**Comment by richsalz** [comment]
Hm, I wonder if those are related to the get_digest API recently added?

**Comment by levitte** [comment]
I can confirm that this is an OpenSSL bug:

``` text
: ; ./util/shlib_wrap.sh apps/openssl crl -noout -fingerprint -in ~/tmp/Net-SSLeay1.80/t/data/verisign.crl.der -inform d
SHA1 Fingerprint=00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
: ; ./util/shlib_wrap.sh apps/openssl version
OpenSSL 1.1.1-dev  xx XXX xxxx
: ; openssl crl -noout -fingerprint -in ~/tmp/Net-SSLeay-1.80/t/data/verisign.crl.der -inform d
SHA1 Fingerprint=2A:49:FA:78:44:40:70:D2:BC:E0:8B:2E:A5:21:30:99:B3:E0:65:FD
: ; openssl version
OpenSSL 1.1.0c  10 Nov 2016
```

**Comment by levitte** [comment]
@richsalz might be on to something...

**Comment by levitte** [comment]
Ah...  when the given md is `EVP_SHA1()`, `X509_CRL_digest()` assumes that the fingerprint has already been pre-computed by the d2i function...  this doesn't seem to be the case here.

``` text
: ; ./util/shlib_wrap.sh gdb apps/openssl --args apps/openssl crl -noout -fingerprint -in ~/tmp/Net-SSLeay-1.80/t/data/verisign.crl.der -inform d
GNU gdb (Debian 7.12-4) 7.12
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from apps/openssl...done.
(gdb) b X509_CRL_digest
Breakpoint 1 at 0x2d0b0
(gdb) r
Starting program: /home/levitte/gitwrk/openssl.net/official/_build/apps/openssl crl -noout -fingerprint -in /home/levitte/tmp/Net-SSLeay-1.80/t/data/verisign.crl.der -inform d
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, X509_CRL_digest (data=0x555555818120, 
    type=0x7ffff7b29140 <sha1_md>, md=0x5555558181d8 "", len=0x0)
    at ../master/crypto/x509/x_all.c:380
380	    if (type == EVP_sha1()) {
(gdb) n
382	        if (len != NULL)
(gdb) n
384	        memcpy(md, data->sha1_hash, sizeof(data->sha1_hash));
(gdb) print *data
$1 = {crl = {version = 0x0, sig_alg = {
      algorithm = 0x7ffff7b2a128 <nid_objs+2600>, parameter = 0x555555818330}, 
    issuer = 0x555555818250, lastUpdate = 0x555555818310, 
    nextUpdate = 0x555555824b10, revoked = 0x555555818350, extensions = 0x0, 
    enc = {
      enc = 0x555555879020 "0\202\342{0\r\006\t*\206H\206\367\r\001\001\005\005", len = 57983, modified = 0}}, sig_alg = {
    algorithm = 0x7ffff7b2a128 <nid_objs+2600>, parameter = 0x5555558872b0}, 
  signature = {length = 256, type = 3, 
    data = 0x5555558872d0 "U\340%\270\277\030\352\366F\251xwT\356[\342u(!\220\234\260\026\240\215\065m\341;rO\331\017\"\025\017\244=O{6\036\357\376R\357\254Z\255Bs\034y\024\214\242}\252@\360\351\357vO\t\224@m\n8\vR\224O\322\033#\214\240\362\261\224\242_\360ʺ\256M\345S\322\016\036$\245\363\v\364{\026\364\322XV\202]\352\237\366E\035\251\216)\003#\356\367j\033!ءr\324\062J\330̔3\262W\351\005\062Ǘ\003Q4\246\264(lhZ\263\062\024%\237\313M\212]\367\376H\334LY2\372\252\253\177\274\016\255:M\271y\265se\016C\024/ea\354\236\307` \a\374[\027_\\4}\265", <incomplete sequence \302>..., flags = 136}, references = 1, flags = 0, akid = 0x0, 
  idp = 0x0, idp_flags = 0, idp_reasons = 32895, crl_number = 0x0, 
  base_crl_number = 0x0, issuers = 0x0, sha1_hash = '\000' <repeats 19 times>, 
  meth = 0x7ffff7b48e00 <int_crl_meth>, meth_data = 0x0, lock = 0x555555818210}
```

**Comment by levitte** [comment]
A quick git grep confirms that `sha1_hash` is never computed.

Ouch.

Also, this shows that we need to improve our CRL test suite.  We're good at verifying the more complex stuff (i.e. to do a cert verification against CRLs), but not so good at the simpler stuff.

**Comment by richsalz** [comment]
So you should compute it and stuff it into the CRL struct and set a flag that it *is* set?

**Comment by kroeckx** [comment]
I was actually assuming this problem, but I couldn't find the commit.

**Comment by kroeckx** [comment]
The code for X509 checks ex_flags, but for the CRL not?

**Comment by levitte** [comment]
Thanks for the pointers, @kroeckx.  PR coming up

**Comment by levitte** [comment]
Fix in #2314.  I haven't written tests for this yet, though.

**Comment by levitte** [comment]
#2314 pushed.  Please try the OpenSSL_1_1_0-stable branch against Net::SSLeay 1.80 and see.
